openedx/core/djangoapps/content/learning_sequences/api/processors/visibility.py

```diff
@@ -25,14 +26,31 @@ def usage_keys_to_remove(self, full_course_outline):
         def should_remove(visibility):
             return visibility.hide_from_toc or visibility.visible_to_staff_only
 
+        can_see_outline = self._is_active_enrolled() or self._is_outline_publicly_visible(full_course_outline)
+
         sections_to_remove = {
             sec.usage_key
             for sec in full_course_outline.sections
-            if should_remove(sec.visibility)
+            if should_remove(sec.visibility) or not can_see_outline
         }
         seqs_to_remove = {
             seq.usage_key
             for seq in full_course_outline.sequences.values()
-            if should_remove(seq.visibility)
+            if should_remove(seq.visibility) or not can_see_outline
         }
         return frozenset(sections_to_remove | seqs_to_remove)
+
```

Comment:
Thinking on this a bit more... at a high level, the stuff you're adding is about enrollment (with anonymous courseware factoring into that because of how enrollment interacts with anonymous access). Maybe it would make sense to create a new OutlineProcessor to encapsulate it (i.e. `EnrollmentOutlineProcessor`)?